src/app/components/ContactEmailSettingsModal.js

```diff
@@ -174,22 +157,47 @@ const ContactEmailSettingsModal = ({ userKeysList, contactID, properties, emailP
     };
 
     useEffect(() => {
+        // prepare the model once mail settings have been loaded
         if (!loadingMailSettings) {
             withLoading(prepare());
         }
     }, [loadingMailSettings]);
 
     useEffect(() => {
-        const updateEncryptToggle = async (keys) => {
-            const expired = await allKeysExpired(keys);
-            if (expired || !keys.length) {
-                setModel((model) => ({ ...model, encrypt: false, keysExpired: expired }));
+        /**
+         * When the list of trusted, expired or revoked keys change,
+         * * update the encrypt toggle (off if all keys are expired or no keys are pinned)
+         * * re-check if the new keys can send
+         * * re-order api keys (trusted take preference)
+         * * move expired keys to the bottom of the list
+         */
+        const noPinnedKeyCanSend = !model.keys.pinned
+            .map((publicKey) => {
+                const fingerprint = publicKey.getFingerprint();
+                const canSend =
+                    !model.expiredFingerprints.has(fingerprint) && !model.revokedFingerprints.has(fingerprint);
+                return canSend;
+            })
+            .filter(Boolean).length;
+        const trustedApiKeys = model.keys.api.filter((key) => model.trustedFingerprints.has(key.getFingerprint()));
+        const noTrustedApiKeyCanSend = !trustedApiKeys.length
```

Comment:
Could we remove this from the `prepare` function? It could be nice just have it one place. Since the dependencies of this `useEffect` should be run after `prepare` we couldn't just leave it here?

Reply:
And it probably doesn't need to be a value we compute in the model like this. What about computing it outside on each render, I don't think it's too heavy? I'd consider moving out `noTrustedApiKeyCanSend` and `noPinnedKeyCanSend`, and then we can change the reduce on line 65 to a for iteration to populate the sets intead.

Reply:
Yeah, that's a better organization. No need to include those variables in the model indeed